stats: add built-in log linear histogram support

ramaraochavali opened the pull request:
*Description*:
Implements built-in log linear histogram. Fixes the issue https://github.com/envoyproxy/envoy/issues/1965

*Risk Level*: High

*Testing*:
Unit, Integration and Manual testing.

*Docs Changes*:
NA

*Release Notes*:

ramaraochavali commented:
@mattklein123 attempted the approach that we talked about in https://github.com/envoyproxy/envoy/issues/1965. Can you PTAL? Still need to 
- Flush the histogram data to sinks
- Correct and add tests

ramaraochavali commented on source/common/stats/thread_local_store.cc:
```diff
@@ -75,6 +91,36 @@ void ThreadLocalStoreImpl::shutdownThreading() {
+void ThreadLocalStoreImpl::mergeHistograms() {
+  if (!shutting_down_) {
+    std::unique_lock<std::mutex> lock(lock_);
+    // TODO: Wait till all threads have been checked-in
```
@mattklein123 I need your help here. How do I wait till all worker threads have finished? Is there a standard pattern in c++/envoy code here that I can use? Tried to look at latch/barrier support but could not find such thing. This is the only thing pending from basic implementation.

mattklein123 replied:
In a perfect world you could add barrier/notify functionality into the TLS itself, e.g., `runOnAllThreadsBarrier(...)`. If you don't want to do that, I would just add a function to TLS to get the number of workers, then in the run on all threads lambda, decrement an atomic counter. Whichever wins the race to 0 can post back to the main thread.

ramaraochavali replied:
Added the method in TLS system.

mattklein123 commented:
@ramaraochavali sorry for the delay I will spend some devoted time to this tomorrow. @mrice32 if you have time it would also be good for you to take a quick look see.

ramaraochavali commented on test/common/stats/thread_local_store_test.cc:
```diff
@@ -101,7 +101,7 @@ TEST_F(StatsThreadLocalStoreTest, NoTls) {
   EXPECT_EQ(&g1, &store_->gauge("g1"));
 
   Histogram& h1 = store_->histogram("h1");
-  EXPECT_EQ(&h1, &store_->histogram("h1"));
```
since we the new approach, if tls is not there we are creating new one instead of returning central ref, this condition fails. I commented it for now and open for discussion on this.

ramaraochavali commented on source/common/stats/grpc_metrics_service_impl.h:
```diff
@@ -135,6 +135,26 @@ class MetricsServiceSink : public Sink {
     gauage_metric->set_value(value);
   }
 
+  void flushHistogram(const Histogram& histogram) override {
```
as discussed, i am just pushing the summary quantiles not rich histogram in the first iteration. If some one is interested, I am open to publishing rich histogram data as well.

ramaraochavali commented:
@trabetti fyi. still WIP. But you can follow this PR

mattklein123 commented on include/envoy/stats/stats.h:
```diff
@@ -194,6 +194,11 @@ class Sink {
+  /**
+   * Flush a histogram.
+   */
+  virtual void flushHistogram(const Histogram&){};
```
PURE (have every sink implement, even if they ignore).

ramaraochavali replied:
yes.. will change them. added them like that so that i can quickly test with out changing tests.

mattklein123 commented on include/envoy/stats/stats.h:
```diff
@@ -263,6 +268,11 @@ class Store : public Scope {
    */
   virtual std::list<GaugeSharedPtr> gauges() const PURE;
+
+  virtual std::list<HistogramSharedPtr> histograms() const {
```
same, PURE, every store should implement.

mattklein123 commented on include/envoy/stats/stats.h:
```diff
@@ -294,6 +304,8 @@ class StoreRoot : public Store {
    */
   virtual void shutdownThreading() PURE;
+
+  virtual void mergeHistograms() {}
```
same: please also add doc comments for all interface functions (above also).

mattklein123 commented on source/common/stats/grpc_metrics_service_impl.h:
```diff
@@ -135,6 +135,26 @@ class MetricsServiceSink : public Sink {
+  void flushHistogram(const Histogram& histogram) override {
+    //  ParentHistogramSharedPtr parent_histogram =
+    //  std::dynamic_pointer_cast<HistogramParentImpl>(&histogram);
+    const HistogramParentImpl& parent_histogram =
```
We shouldn't dynamic cast here. Add whatever interface functions you need.

mattklein123 commented on source/common/stats/thread_local_store.h:
```diff
@@ -74,6 +75,10 @@ class ThreadLocalStoreImpl : public StoreRoot {
 
+  void mergeHistograms() override;
+
+  void mergeInternal();
```
private

mattklein123 commented on source/server/http/admin.cc:
```diff
@@ -400,11 +402,24 @@ Http::Code AdminImpl::handlerStats(absl::string_view url, Http::HeaderMap& respo
 
+  for (const Stats::HistogramSharedPtr& histogram : server_.stats().histograms()) {
+    Stats::ParentHistogramSharedPtr parent_hist_ptr =
+        std::dynamic_pointer_cast<Stats::HistogramParentImpl>(histogram);
```
no dynamic cast

mattklein123 commented on source/common/stats/thread_local_store.cc:
```diff
@@ -215,18 +261,22 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
+  HistogramSharedPtr hist_tls_ptr;
+  hist_tls_ptr.reset(new ThreadLocalHistogramImpl(final_name, parent_,
+                                                  std::move(tag_extracted_name), std::move(tags)));
+  dynamic_cast<HistogramParentImpl&>(*central_ref).addTlsHistogram(hist_tls_ptr);
```
no dynamic cast. You can store a local ref/pointer if needed to have access to the impl, or just the local map to storing impls.

mattklein123 commented on source/common/stats/thread_local_store.cc:
```diff
@@ -215,18 +261,22 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
+                                              std::move(tags)));
   }
+  HistogramSharedPtr hist_tls_ptr;
+  hist_tls_ptr.reset(new ThreadLocalHistogramImpl(final_name, parent_,
```
nit: merge into previous line, or do directly in the context of addTlsHistogram()

mattklein123 commented on source/common/stats/thread_local_store.cc:
```diff
@@ -75,6 +91,36 @@ void ThreadLocalStoreImpl::shutdownThreading() {
+  for (ScopeImpl* scope : scopes_) {
+    for (auto histogram : scope->central_cache_.histograms_) {
+      ParentHistogramSharedPtr parent_hist_ptr =
+          std::dynamic_pointer_cast<HistogramParentImpl>(histogram.second);
```
to remove all the dynamic casts here, I would change the map to store impls so you can use them locally.

mattklein123 commented on source/common/stats/thread_local_store.cc:
```diff
@@ -75,6 +91,36 @@ void ThreadLocalStoreImpl::shutdownThreading() {
+
+void ThreadLocalStoreImpl::mergeInternal() {
+  std::unique_lock<std::mutex> lock(lock_);
+  for (ScopeImpl* scope : scopes_) {
```
I would use `histograms()` here, which will acquire the lock and quickly sort through histograms that need merging. Then, you can operate on each individual histogram safely without any lock since the merging process will take a larger amount of time.

mattklein123 commented on source/common/stats/thread_local_store.cc:
```diff
@@ -75,6 +91,36 @@ void ThreadLocalStoreImpl::shutdownThreading() {
+      for (ScopeImpl* scope : scopes_) {
+        for (auto histogram : tls_->getTyped<TlsCache>().scope_cache_[scope].histograms_) {
+          TlsHistogramSharedPtr tls_histogram_ptr =
+              std::dynamic_pointer_cast<ThreadLocalHistogramImpl>(histogram.second);
```
remove dynamic_cast (by storing impls if needed). I will stop commenting but basically the change should have no dynamic_casts.

mattklein123 commented on source/common/stats/thread_local_store.cc:
```diff
@@ -75,6 +91,36 @@ void ThreadLocalStoreImpl::shutdownThreading() {
+        }
+      }
+    });
+    if (main_thread_dispatcher_) {
```
I think main_thread_dispatcher_ must be non-nullptr here. I would skip this check? Obviously this also needs to get moved inside the run on all threads lambda above, see my comment above.

mattklein123 commented on source/server/server.cc:
```diff
@@ -139,6 +145,7 @@ void InstanceImpl::flushStats() {
   server_stats_->days_until_first_cert_expiring_.set(
       sslContextManager().daysUntilFirstCertExpires());
 
+  stats_store_.mergeHistograms();
```
If we do this, the merge will essentially always be one interval behind the flush interval. Is this what we want? Do we want to wait into the merge is complete? If so, my thinking was to have mergeHistograms() take a lambda which would fire when the merge is complete, and then the flush would complete.

Also, on this topic, for thought: I do wonder if the histogram merge interval should be separately configurable from the main stat flush interval. (E.g., let's say you wanted to flush stats/gauges every 500ms, you might want to wait longer to get more histogram data per interval). I would put in a TODO to consider this in the future.

ramaraochavali replied:
I did not get the first part. Since we block till all tls histograms are merged then main merge goes in sequentially - is the lambda still required?

mattklein123 commented on source/server/server.cc:
```diff
@@ -139,6 +145,7 @@ void InstanceImpl::flushStats() {
       sslContextManager().daysUntilFirstCertExpires());
 
+  stats_store_.mergeHistograms();
   InstanceUtil::flushCountersAndGaugesToSinks(config_->statsSinks(), stats_store_);
```
Can you rename this function to something more appropriate? `flushMetricsToSinks()` ?

jmarantz commented on source/common/stats/grpc_metrics_service_impl.h:
```diff
@@ -135,6 +135,26 @@ class MetricsServiceSink : public Sink {
+        dynamic_cast<const HistogramParentImpl&>(histogram);
+    io::prometheus::client::MetricFamily* metrics_family = message_.add_envoy_metrics();
+    metrics_family->set_type(io::prometheus::client::MetricType::SUMMARY);
+    metrics_family->set_name(parent_histogram.name());
```
feels like this should go into the .cc

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -23,6 +23,10 @@
 #include "absl/strings/string_view.h"
 
+extern "C" {
+#include <circllhist.h>
```
#include "circlehist.h"

<> should be used for system includes (e.g. included with the OS or compiler). Also I would've thought this should be in a subdir.

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -369,6 +373,126 @@ class HistogramImpl : public Histogram, public MetricImpl {
+struct HistogramStatistics {
+public:
+  HistogramStatistics(histogram_t* histogram_ptr) {
+    double quantile_in[] = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
```
factor out this constant so it only appears once.  I see it also in grpc_metrics_service_impl.h:149

ramaraochavali replied:
i am not able to make it const - because of my c++ knowledge levels :-). But made it class variable and used it every where.

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -369,6 +373,126 @@ class HistogramImpl : public Histogram, public MetricImpl {
+                       quantile_out_[4], quantile_out_[5], quantile_out_[6], quantile_out_[7],
+                       quantile_out_[8]);
+  }
+  double quantile_out_[9];
```
double quantile_out_[ARRAY_SIZE(quantile_in)];

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -369,6 +373,126 @@ class HistogramImpl : public Histogram, public MetricImpl {
+  }
+
+  void beginMerge() {
+    if (current_active_ == 0) {
```
suggestion: current_active_ = 1 - current_active_;

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -369,6 +373,126 @@ class HistogramImpl : public Histogram, public MetricImpl {
+
+  void merge(histogram_t* target) {
+    histogram_t* hist_array[1];
+    if (current_active_ == 0) {
```
suggestion: hist_array[0] = histograms_[1 - current_active_];

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -369,6 +373,126 @@ class HistogramImpl : public Histogram, public MetricImpl {
+    } else {
+      hist_array[0] = histograms_[0];
+    }
+    hist_accumulate(target, hist_array, 1);
```
s/1/ARRAY_SIZE(hist_array)/

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -369,6 +373,126 @@ class HistogramImpl : public Histogram, public MetricImpl {
+  }
+
+  Store& parent_;
+  std::list<HistogramSharedPtr> tls_histograms_;
```
could we use the more specific type here to avoid dynamic_cast?

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -369,6 +373,126 @@ class HistogramImpl : public Histogram, public MetricImpl {
+  void addTlsHistogram(HistogramSharedPtr hist_ptr) { tls_histograms_.emplace_back(hist_ptr); }
+
+  void merge() {
+    hist_clear(interval_histogram_);
```
could you comment about the threading model here?  or maybe just point to Matt's design in github?

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -369,6 +373,126 @@ class HistogramImpl : public Histogram, public MetricImpl {
+  }
+
+  HistogramStatistics getIntervalHistogramStatistics() const {
+    HistogramStatistics interval_histogram(interval_histogram_);
```
suggest: return HistogramStatistics(interval_histogram_);
...just for brevity

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -369,6 +373,126 @@ class HistogramImpl : public Histogram, public MetricImpl {
+  }
+
+  HistogramStatistics getCumulativieHistogramStatistics() const {
+    HistogramStatistics cumulative_histogram(cumulative_histogram_);
```
ditto.

jmarantz commented on source/common/stats/thread_local_store.cc:
```diff
@@ -61,6 +61,22 @@ std::list<GaugeSharedPtr> ThreadLocalStoreImpl::gauges() const {
+  for (ScopeImpl* scope : scopes_) {
+    for (auto histogram : scope->central_cache_.histograms_) {
+      if (names.insert(histogram.first).second) {
+        ret.push_back(histogram.second);
```
probably worth a note here that the histograms will come back in arbitrary order.

suggestion: return a vector rather than a list, and build that vector after accumulating the set, so that:
  a) you can let go of the lock before converting to a vector
  b) you can do vector.reserve(names_.size()) before building it.

ramaraochavali replied:
@jmarantz have not addressed this yet. will address in next round. :-)

jmarantz replied:
I realize on re-read that there's an established pattern of using lists rather than for returning collections of stats.  @mattklein123 -- I'm curious why that choice?

But for this PR I retract this suggestion -- it probably doesn't save that much to release the lock earlier.  I would like to see some comments as to why this de-duping is necessary though; that seems like something must have gone wrong upstream for it to be needed.

mattklein123 replied:
I don't remember why this was a list. I think this was historical. Free free to return a vector if you want to change the various usages. 

re: The de-dup, that is required due to the threaded store uses scopes.

jmarantz replied:
Shouldn't we incorporate the scopes into the histogram names somehow, so we have distinct ones for each scope?  It seems like just dropping them would lose data.

mattklein123 replied:
Hmm yes, this is a problem for histograms which don't use shared mem backing store. It's not an issue for counters/gauges. Good catch. This will need to be fixed somehow.

jmarantz replied:
Cool... @ramaraochavali you can maybe leave a TODO here, and maybe even a warning log that we are dropping content?

jmarantz commented on source/common/stats/thread_local_store.cc:
```diff
@@ -61,6 +61,22 @@ std::list<GaugeSharedPtr> ThreadLocalStoreImpl::gauges() const {
+  std::unordered_set<std::string> names;
+  std::unique_lock<std::mutex> lock(lock_);
+  for (ScopeImpl* scope : scopes_) {
+    for (auto histogram : scope->central_cache_.histograms_) {
```
The style-guide (https://google.github.io/styleguide/cppguide.html#auto) leaves some room for interpretation, but my reading is that in this case and a few others in this PR, you should use explicit types.

If you were using iterator objects, those are good candidates for 'auto' because their types are fairly obvious and not necessarily important to spell out.

e..g.
```
for (auto it = collection.begin(); it != collection.end(); ++it) {
  explicit_type& val = *it;
  ...
};
```
or
```
for (explicit_type& val : collection) {
  ...
};
```
But it's possible there's an established style here I'm going against the grain in suggesting.

ramaraochavali replied:
I am changing it like this
  for (ScopeImpl* scope : scopes_) {
    for (auto histogram : scope->central_cache_.histograms_) {
      const std::string& hist_name = histogram.first;
      const ParentHistogramSharedPtr& parent_hist = histogram.second;
      if (names.insert(hist_name).second) {
        ret.push_back(parent_hist);
      }
    }
  }

ramaraochavali commented:
@mattklein123 ready for another pass. The new method `runOnAllThreadsWithBarrier` seems to be violating asan and tsan tests while decrementing the atomic integer. Will look in to that tomorrow. Let me know if you have ideas around that.

mattklein123 commented on bazel/external/libcircllhist.BUILD:
```diff
@@ -0,0 +1,17 @@
+
+genrule(
+    name = "circllhist_config",
+    srcs = ["src/circllhist_config.h.in"],
```
Can we follow up on @jmillikin-stripe comment here? It does seem weird that the generated header is just being copied. Can you add a comment as to why this is OK if it is? cc @postwait

postwait replied:
The _config.h.in is an autoconf input template.  It is not safe to copy it on all platforms.  There is a
```
#ifdef HAVE_ALLOCA_H
```

in the circllhist.c file that needs that that conditionally set depending on system headers.

ramaraochavali replied:
@postwait  thanks. What is your suggestion here? Should I copy based on platform? or which platforms it is good to copy?

postwait replied:
libcircllhist (like many open source projects) is auto-tools based.  When you build it, it will build-time detect which features are available on your platform and act on that.  I don't know how to do the same thing in cmake.  I am pretty sure that HAVE_ALLOCA_H is the only important thing in that header though -- so if you can get cmake to provide that, you're set.

postwait replied:
Looking at this, it is only alloca() that is a problem.  I've opened a PR on libcircllhist to resolve this.  It should be possible shortly to just ignore that header completely.

ramaraochavali replied:
@postwait thank you so much.

ramaraochavali replied:
@postwait Awesome. Thanks. I got the new version and removed the genRule and it works.

mattklein123 commented on include/envoy/stats/stats.h:
```diff
@@ -167,6 +172,21 @@ class Histogram : public virtual Metric {
+  /**
+   * Returns the Histogram Summary Statistics for the flush interval.
+   */
+  virtual HistogramStatistics intervalStatistics() const PURE;
```
I'm not actually sure how this compiles. I think it's luck if `HistogramStatistics` is returned by value. We either need to define `HistogramStatistics` in this file, or I think probably more optimally, create an abstract interface for `HistogramStatistics` and return by const ref. I'm not sure which one will be easier. (The question on whether we can return by const ref is not really dependent on whether we use an abstract interface or not.)

mattklein123 commented on include/envoy/stats/stats.h:
```diff
@@ -167,6 +172,21 @@ class Histogram : public virtual Metric {
+  virtual void merge() PURE;
+
+  /**
+   * Returns the Histogram Summary Statistics for the flush interval.
```
super nit: don't capitalize "Histogram Summary Statistics"

mattklein123 commented on include/envoy/stats/stats.h:
```diff
@@ -167,6 +172,21 @@ class Histogram : public virtual Metric {
+  virtual HistogramStatistics intervalStatistics() const PURE;
+
+  /**
+   * Returns the Cumulative Histogram Summary Statistics.
```
super nit: don't capitalize "Cumulative Histogram Summary Statistics"

mattklein123 commented on include/envoy/stats/stats.h:
```diff
@@ -194,6 +214,11 @@ class Sink {
+  /**
+   * Flush a histogram.
+   */
+  virtual void flushHistogram(const Histogram&) PURE;
```
nit: `const Histogram& histogram`

mattklein123 commented on include/envoy/thread_local/thread_local.h:
```diff
@@ -46,6 +46,13 @@ class Slot {
+  /**
+   * Run a callback on all registered threads with a barrier.
+   * @param cb supplies the callback to run on each thread.
+   * @param main_callback supplies the callback to run on main thread after threads are done.
```
Can you add a warning here that there is no guarantee that this callback will fire during shutdown. I think that's OK for now but I would like to clearly mark that in case we need that later.

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -23,6 +23,10 @@
 #include "absl/strings/string_view.h"
 
+extern "C" {
+#include "circllhist.h"
```
Can you add a TODO here to get extern "C" in the headers for the library? (Or just do a PR over there). cc @postwait

postwait replied:
Fixed upstream.

ramaraochavali replied:
@postwait awesome. I updated the latest version and it is no longer needed.

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -158,6 +162,12 @@ class Utility {
+/**
+ * Flags used by all stats types to figure out whether they have been used.
+ */
+struct Flags {
```
Can we put this inside MetricsImpl?

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -333,27 +340,50 @@ class GaugeImpl : public Gauge, public MetricImpl {
+                       quantiles_out_[8]);
+  }
+
+  double quantiles_in_[9] = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
```
These should all be static constants. If you can't figure out how to do the C++ I think either @mrice32 or @jmarantz can hopefully help.

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -333,27 +340,50 @@ class GaugeImpl : public Gauge, public MetricImpl {
+ */
+struct HistogramStatistics {
+public:
+  HistogramStatistics() {}
```
I'm confused as to what the default constructor here does. I think at minimum you need to zero out the array?

ramaraochavali replied:
I have added the default constructor to return empty HistogramStatistics for Histograms that do not need this functionality - are you suggesting initialize the array with zeros here?

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -366,9 +396,135 @@ class HistogramImpl : public Histogram, public MetricImpl {
+      : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent) {
+    histograms_[0] = hist_alloc();
+    histograms_[1] = hist_alloc();
+    current_active_ = 0;
```
initialize this and next line in initializer list

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -366,9 +396,135 @@ class HistogramImpl : public Histogram, public MetricImpl {
+
+  bool used() const override { return flags_ & Flags::Used; }
+
+  HistogramStatistics intervalStatistics() const override { return HistogramStatistics(); }
```
Still confused here about how we are returning the summary data...

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -366,9 +396,135 @@ class HistogramImpl : public Histogram, public MetricImpl {
+
+  Store& parent_;
+  std::list<TlsHistogramSharedPtr> tls_histograms_;
+  histogram_t* interval_histogram_;
```
Yeah, I think in looking at this now I would make HistogramStatistics an interface, then where you need it, have an impl that wraps the circl implementation, which you can then return via const reference.

mattklein123 commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -91,6 +91,20 @@ void InstanceImpl::runOnAllThreads(Event::PostCb cb) {
+void InstanceImpl::runOnAllThreadsWithBarrier(Event::PostCb cb, Event::PostCb main_callback) {
+  ASSERT(std::this_thread::get_id() == main_thread_id_);
+  ASSERT(!shutdown_);
+  worker_counter_ = registered_threads_.size();
```
This is not safe for concurrent uses of this function. Instead of using a member variable, allocate an atomic inside a shared_ptr, and capture it in the posted lambdas. 

Also, this new function needs explicit tests in thread_local_impl_test

ramaraochavali replied:
would it be used concurrently since it is allowed to be called only from main thread? Any way I will try with shared_ptr as it seems better compared to member variable.

ramaraochavali replied:
I have changed it to use shared_ptr. I came across this http://floating.io/2017/07/lambda-shared_ptr-memory-leak/ - Should we worry about this?

ramaraochavali replied:
reg tests, I will add once the implementation comes to a sane stage.

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -366,9 +396,135 @@ class HistogramImpl : public Histogram, public MetricImpl {
 };
 
+/**
+ * Log Linear Histogram implementation per thread.
```
Can we move  ThreadLocalHistogramImpl and HistogramParentImpl directly into thread_local_store. We are going to have to deal with some locking issues and I would like things to be more explicitly coupled. We can break it apart later if necessary.

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -366,9 +396,135 @@ class HistogramImpl : public Histogram, public MetricImpl {
+
+  bool used() const override {
+    bool any_tls_used = false;
+    for (auto tls_histogram : tls_histograms_) {
```
All sues of tls_histograms_ have to be locked if we are going to do the used() stuff. I think what I would actually do is move your "merge_lock" into each histogram. and then just lock each function here that is needed. Let's go ahead and do that and I can review again. (Just trying to iteratively tighten stuff up so I can focus more on concurrency).

ramaraochavali replied:
ok. Moved the merge_lock to histogram and locked the functions there. Looks like this is much cleaner.

ramaraochavali commented:
@mattklein123 ready for another pass. Only comments not yet addressed are
 1. Bazel rule for auto processing the header and 
 2. making double array const in HistogramStatistics. 
Will address them soon.

postwait commented:
https://github.com/circonus-labs/libcircllhist/pull/23 will remove the need for processing that header.  It is under review now.

ramaraochavali commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,142 @@
+  }
+  // Stats::Histogram
+  void recordValue(uint64_t value) override {
+    std::unique_lock<std::mutex> lock(merge_lock_);
```
here lock may not be required. need to think through a bit here.

mattklein123 replied:
Lock is not needed here. Because of the RCU locking style we are using it's always safe to directly hit the current histogram.

postwait commented:
That has been put back, circllhist.c doesn't need that header anymore. It has since been removed.

mattklein123 commented on include/envoy/stats/stats.h:
```diff
@@ -146,12 +150,27 @@ class Gauge : public virtual Metric {
+   */
+  virtual std::string summary() const PURE;
+
+  double quantiles_in_[9] = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
```
These are an implementation detail right now, right? Can we move this out of the interface and into the implementation?

ramaraochavali replied:
i am using these in MetricsService otherwise I have to use dynamic_cast to impl and access.

jmarantz replied:
Maybe use a pure method?

  double* quantiles() const PURE;
  size_t numQuantiles() const PURE;

or you could return std::pair<double*, size_t> -- not sure which is preferred.

mattklein123 replied:
+1, what about just returning a `const std::vector<double>&` via a pure virtual method? This can be hidden by the implementation.

ramaraochavali replied:
changed it to use a pure virtual method returning const std::vector<double>&

mattklein123 commented on include/envoy/thread_local/thread_local.h:
```diff
@@ -46,6 +46,14 @@ class Slot {
   virtual void runOnAllThreads(Event::PostCb cb) PURE;
 
+  /**
+   * Run a callback on all registered threads with a barrier. There is no guarantee that this will
```
nit: "that the callback will be..."

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -277,6 +275,13 @@ struct RawStatData {
+  /**
+   * Flags used by all stats types to figure out whether they have been used.
+   */
+  struct Flags {
```
nit: I think that can be protected?

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -333,40 +338,77 @@ class GaugeImpl : public Gauge, public MetricImpl {
+ */
+class HistogramStatisticsImpl : public HistogramStatistics {
+public:
+  HistogramStatisticsImpl() {}
```
Per my last comment, I'm pretty sure you have to zero out the storage array here. Otherwise won't summary() return bogus data?

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -333,40 +338,77 @@ class GaugeImpl : public Gauge, public MetricImpl {
                 std::vector<Tag>&& tags)
-      : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent) {}
+      : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent) {
+    cumulative_statistics_.reset(new HistogramStatisticsImpl());
```
init in initializer list. Do these need to be pointers at all? Can't we just allocate them directly as members?

mattklein123 commented on source/common/stats/thread_local_store.cc:
```diff
@@ -214,19 +252,22 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
+                                              std::move(tags)));
   }
+  TlsHistogramSharedPtr hist_tls_ptr(new ThreadLocalHistogramImpl(
+      final_name, parent_, std::move(tag_extracted_name), std::move(tags)));
```
you can't move these things twice. You will need to copy in one case. I would copy in the central ref case as that is less common. (I think it's time to start adding tests, you will uncover some of these issues.)

mattklein123 commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,142 @@
+        current_active_(0), flags_(0) {
+    histograms_[0] = hist_alloc();
+    histograms_[1] = hist_alloc();
+    cumulative_statistics_.reset(new HistogramStatisticsImpl());
```
Same comment as elsewhere. Does this need to be heap allocated? Can't it just be a member directly?

mattklein123 commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,142 @@
+  }
+
+  void merge(histogram_t* target) {
+    std::unique_lock<std::mutex> lock(merge_lock_);
```
lock not needed here either.

mattklein123 commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,142 @@
+  std::atomic<uint16_t> flags_;
+  std::shared_ptr<HistogramStatistics> interval_statistics_;
+  std::shared_ptr<HistogramStatistics> cumulative_statistics_;
+  mutable std::mutex merge_lock_;
```
I think you don't need any lock for the thread local variant.

mattklein123 commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,142 @@
+      : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent) {
+    interval_histogram_ = hist_alloc();
+    cumulative_histogram_ = hist_alloc();
+    cumulative_statistics_.reset(new HistogramStatisticsImpl(cumulative_histogram_));
```
same comment

mattklein123 commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,142 @@
+    return any_tls_used;
+  }
+
+  void addTlsHistogram(TlsHistogramSharedPtr hist_ptr) { tls_histograms_.emplace_back(hist_ptr); }
```
lock

mattklein123 commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,142 @@
+  void recordValue(uint64_t) override {}
+
+  bool used() const override {
+    bool any_tls_used = false;
```
lock

mattklein123 commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,142 @@
+   * https://github.com/envoyproxy/envoy/issues/1965#issuecomment-376672282.
+   */
+  void merge() override {
+    if (used()) {
```
would be nice to have a `usedWorker()` which asserts lock is already acquired, then you can acquire lock once at top

ramaraochavali replied:
I did not get this comment. Can you elaborate more?

mattklein123 replied:
My comment here is that you acquire the lock for `used()`, then release it, then reacquire it. A better model would be to do something like:

```
bool used() override {
  <lock>
  usedWorker();
}

void merge() override {
  <lock>
  if (usedWorker()) {
    ..
  }
}
```

You can avoid double acquiring

mattklein123 commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -91,6 +91,21 @@ void InstanceImpl::runOnAllThreads(Event::PostCb cb) {
+void InstanceImpl::runOnAllThreadsWithBarrier(Event::PostCb cb, Event::PostCb main_callback) {
+  ASSERT(std::this_thread::get_id() == main_thread_id_);
+  ASSERT(!shutdown_);
+  std::shared_ptr<std::atomic<int>> worker_count =
```
int -> uint64_t

mattklein123 commented on source/server/server.cc:
```diff
@@ -138,9 +146,12 @@ void InstanceImpl::flushStats() {
-  InstanceUtil::flushCountersAndGaugesToSinks(config_->statsSinks(), stats_store_);
-  stat_flush_timer_->enableTimer(config_->statsFlushInterval());
+  // TODO(ramaraochavali): consider adding different flush interval for histograms.
+  if (!terminated_) {
```
why !terminated_ here? Does it matter? I would remove this.

mattklein123 commented on source/server/server.cc:
```diff
@@ -138,9 +146,12 @@ void InstanceImpl::flushStats() {
-  stat_flush_timer_->enableTimer(config_->statsFlushInterval());
+  // TODO(ramaraochavali): consider adding different flush interval for histograms.
+  if (!terminated_) {
+    stats_store_.mergeHistograms();
```
Same comment applies, this is going to be one cycle behind since this will not block and the post back will come after this runs. If we don't care about this, please add some comments to this effect. I still think megeHistograms() should take a lambda which is invoked when the merge is complete.

ramaraochavali replied:
I agree that sending metrics one cycle behind is not good. I have added the lambda to mergeHistograms which will complete the final stage of stats flush

jmarantz commented on include/envoy/thread_local/thread_local.h:
```diff
@@ -46,6 +46,14 @@ class Slot {
 
+  /**
+   * Run a callback on all registered threads with a barrier. There is no guarantee that the
+   * callback be fired during shutdown.
```
Again really just to clarify semantics, would it be correct to rephrase this:

A shutdown initiated during the running of the PostCBs may prevent main_callback from being called.

WDYT?

mattklein123 replied:
Sure

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -333,27 +342,51 @@ class GaugeImpl : public Gauge, public MetricImpl {
+  }
+
+  std::string summary() const override {
+    return fmt::format("P0: {} , P25: {}, P50: {}, P75: {}, P90: {}, P95: {}, P99: {}, P100: {}",
```
WDYT of constructing this by looping over the quantiles_in_ and quantiles_out_ vectors, so they'll never get out of sink. In fact it looks like they are out of sink now as you have a .999 in the quantiles_in_ initializer above; only 8 values in your format string and 9 args below.

Easy to miss.  Let's avoid that by constructing this from the source of truth.

mattklein123 replied:
+1

jmarantz commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,139 @@
+    if (used()) {
+      std::unique_lock<std::mutex> lock(merge_lock_);
+      hist_clear(interval_histogram_);
+      for (auto tls_histogram : tls_histograms_) {
```
again, would prefer explicit types here.

ramaraochavali commented:
@jmarantz  I have already implemented it to render on the admin console, /stats end point. See admin.cc changes.

jmarantz commented:
@ramaraochavali nice!

mattklein123 commented on include/envoy/stats/stats.h:
```diff
@@ -294,6 +344,12 @@ class StoreRoot : public Store {
   virtual void shutdownThreading() PURE;
+
+  /**
+   * Called during the flush process to merge all the thread local histograms. The passed in
```
The callback will be called on the main thread, but it will happen after the method returns. Please clarify. :)

ramaraochavali replied:
clarified. lmk any changes are needed here.

mattklein123 commented on include/envoy/stats/stats.h:
```diff
@@ -263,10 +303,20 @@ class Store : public Scope {
+/**
+ * Callback invoked when a store's mergeHistogram() runs.
+ */
+typedef std::function<void()> PostMergeCb;
```
nit: I think we can probably kill the typedef for this, just use std::function inline below.

ramaraochavali replied:
I have added it because, it is repeated in tests as well. if you are strongly think about removing it, I can remove it.

mattklein123 commented on include/envoy/stats/stats.h:
```diff
@@ -294,6 +344,12 @@ class StoreRoot : public Store {
+   * Called during the flush process to merge all the thread local histograms. The passed in
+   * callback will be invoked after the merge process is complete.
+   */
+  virtual void mergeHistograms(PostMergeCb mergeCb) PURE;
```
nit: s/mergeCb/merge_complete_cb

mattklein123 commented on include/envoy/thread_local/thread_local.h:
```diff
@@ -46,6 +46,14 @@ class Slot {
+   * Run a callback on all registered threads with a barrier. There is no guarantee that the
+   * callback be fired during shutdown.
+   * @param cb supplies the callback to run on each thread.
+   * @param main_callback supplies the callback to run on main thread after threads are done.
```
s/main_callback/all_threads_complete_cb (or something). Also per above clarify this runs on the main thread. You should also ASSERT in the impl that this function is only called on the main thread.

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -280,10 +278,21 @@ class MetricImpl : public virtual Metric {
   MetricImpl(const std::string& name, std::string&& tag_extracted_name, std::vector<Tag>&& tags)
       : name_(name), tag_extracted_name_(std::move(tag_extracted_name)), tags_(std::move(tags)) {}
 
+  MetricImpl(const std::string& name, std::string& tag_extracted_name, std::vector<Tag>& tags)
```
const for the last 2 params

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -333,27 +342,51 @@ class GaugeImpl : public Gauge, public MetricImpl {
+class HistogramStatisticsImpl : public HistogramStatistics {
+public:
+  HistogramStatisticsImpl() {
+    for (size_t i = 0; i < ARRAY_SIZE(quantiles_out_); i++) {
```
It's not a big deal either way but you can also just do memset here on the array, though, in general, for ease of programming I would suggest using std::array at minimum and maybe std::vector.

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -366,7 +399,21 @@ class HistogramImpl : public Histogram, public MetricImpl {
   // Stats::Histogram
   void recordValue(uint64_t value) override { parent_.deliverHistogramToSinks(*this, value); }
 
+  void merge() override {}
```
Does this ever get called? If not I would do `NOT_IMPLEMENTED`. Same for used(). It would probably be better to actually split the histogram interfaces into two different interfaces to avoid this. (Basically some type of parent interface, and then a basic histogram interface). If you don't want to do that I would do NOT_IMPLEMENTED and put in a TODO.

mattklein123 commented on source/common/stats/thread_local_store.cc:
```diff
@@ -75,6 +91,29 @@ void ThreadLocalStoreImpl::shutdownThreading() {
+          }
+        },
+        [this]() -> void { mergeInternal(); });
+    mergeCb();
```
You want to pass the callback here, not the invocation of the callback. In general at this point I think this is the last review cycle I'm going to do until there is full test coverage, because some of these issues would be really easy to spot with tests.

mattklein123 commented on source/common/stats/thread_local_store.cc:
```diff
@@ -214,19 +253,21 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
-    std::string tag_extracted_name = parent_.getTagsForName(final_name, tags);
-    central_ref.reset(
-        new HistogramImpl(final_name, parent_, std::move(tag_extracted_name), std::move(tags)));
+    central_ref.reset(new HistogramParentImpl(final_name, parent_, tag_extracted_name, tags));
```
I actually am not sure that the compiler will not pick the move version here. Given that this is hazy C++ area, let's avoid it. I would just go back to a single constructor version with move, and do an explicit copy in here with a comment.

mattklein123 commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,139 @@
+  Store& parent_;
+
+private:
+  int current_active_;
```
uint64_t

mattklein123 commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,139 @@
+
+  void beginMerge() { current_active_ = 1 - current_active_; }
+
+  void merge() override {}
```
Same NOT_IMPLEMENTED comment, prefer interface split.

mattklein123 commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,139 @@
+  }
+
+  // Stats::Histogram
+  void recordValue(uint64_t) override {}
```
NOT_IMPLEMENTED/interface split

mattklein123 commented on source/common/stats/thread_local_store.h:
```diff
@@ -126,6 +269,5 @@ class ThreadLocalStoreImpl : public StoreRoot {
   Counter& num_last_resort_stats_;
   HeapRawStatDataAllocator heap_allocator_;
 };
-
```
super nit: remove this newline deletion

mattklein123 commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -91,6 +91,21 @@ void InstanceImpl::runOnAllThreads(Event::PostCb cb) {
 }
 
+void InstanceImpl::runOnAllThreadsWithBarrier(Event::PostCb cb, Event::PostCb main_callback) {
+  ASSERT(std::this_thread::get_id() == main_thread_id_);
```
disregard other comment about ASSERT, you already have it!

mattklein123 commented on source/server/server.cc:
```diff
@@ -140,8 +148,11 @@ void InstanceImpl::flushStats() {
-  stat_flush_timer_->enableTimer(config_->statsFlushInterval());
+  // TODO(ramaraochavali): consider adding different flush interval for histograms.
+  stats_store_.mergeHistograms([this]() -> void {
+    InstanceUtil::flushMetricsToSinks(config_->statsSinks(), stats_store_);
```
You should do the rest of flushStats() above in here.

ramaraochavali commented:
@jmarantz addressed your comments about using vector. PTAL.

ramaraochavali commented on test/common/stats/thread_local_store_test.cc:
```diff
@@ -190,6 +190,182 @@ TEST_F(StatsThreadLocalStoreTest, BasicScope) {
+                                          "P95: 2.095, P99: 2.099, P99.9: 2.0999, P100: 2.1";
+  std::string not_used_summary =
+      "P0: 0, P25: 0, P50: 0, P75: 0, P90: 0, P95: 0, P99: 0, P99.9: 0, P100: 0";
+  std::string not_used_in_interval_summary =
```
@postwait after we call `hist_free` of an existing histogram and get quantiles on it we get all 1's instead of all 0's as we get with newly allocated histogram. Is that expected?

postwait replied:
You can't call quantiles on a histogram you've freed, that would crash or quantile trash.  Use `hist_clear` to remove all samples from a histogram (but leave the bins in place).  If you call the quantile function on a histogram and get 1's it is an indicator that it has value(s) that exist in the [1.0,1.1) bin.

ramaraochavali replied:
Sorry. I meant hist_clear only. After I call hist_clear I am getting all 1’s.

ramaraochavali replied:
i am using hist_alloc(). If i call hist_free and hist_alloc, it returns 0's but after hist_clear return 1's
Sample code that reproduces the problem
  histograms_[0] = hist_alloc();
  hist_insert_intscale(histograms_[0], 1, 0, 1);
  hist_clear(histograms_[0]);

jmarantz replied:
Try --config=asan or running under valgrind.

jmarantz replied:
Try --config=asan or running under valgrind.

jmarantz replied:
Try --config=asan or running under valgrind.

jmarantz replied:
Try --config=asan or running under valgrind.

postwait replied:
Test and fix added upstream.  Nice catch.  Edge case on empty histograms for quantiles.

Your output double vector will be full of NaN's now if you ask about an empty histogram.

ramaraochavali replied:
@postwait thanks a lot. I updated the new version and it works correctly.

ramaraochavali replied:
@postwait looks like found another corner case issue or may be i am misunderstanding.

Consider the code

    histogram_t* main_thread_interval_hist = hist_alloc();
    histogram_t* per_thread_interval_hist = hist_alloc();
    hist_insert_intscale(per_thread_interval_hist, 1, 0, 1);
    double in[9] = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
    double out[9];
    hist_approx_quantile(main_thread_interval_hist, in, 9,out);
    histogram_t* hist_array[1];
    hist_array[0] = per_thread_interval_hist;
    hist_accumulate(main_thread_interval_hist, hist_array, ARRAY_SIZE(hist_array));
    hist_clear(per_thread_interval_hist);
    //hist_free(per_thread_interval_hist);
    //per_thread_interval_hist = hist_alloc();
    hist_array[0] = per_thread_interval_hist;
    hist_insert_intscale(per_thread_interval_hist, 2, 0, 1);
    //hist_clear(main_thread_interval_hist);
    //hist_free(main_thread_interval_hist);
    main_thread_interval_hist = hist_alloc();
    hist_accumulate(main_thread_interval_hist, hist_array, ARRAY_SIZE(hist_array));
    hist_approx_quantile(main_thread_interval_hist, in, 9,out);
    std::cout<<"after accumulations :"<<out[0]<<"==> \n";

    histogram_t* direct_hist = hist_alloc();
    hist_insert_intscale(direct_hist, 2, 0, 1);
    hist_approx_quantile(direct_hist, in, 9,out);
    std::cout<<"direct :"<<out[0]<<"==> \n";

I expect direct and after accumulations to give the same output. But it gives 1 in first bin (rest of the bins are same) for accumulation case and 2 in the direct case.
If I change the code to free and hist_alloc it works correctly in both the cases. Is this is a corner case issue with hist_clear?

ramaraochavali replied:
@postwait Please see the above comment

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -333,27 +339,62 @@ class GaugeImpl : public Gauge, public MetricImpl {
+  const std::vector<double>& computedQuantiles() const override { return computed_quantiles_; }
+
+private:
+  std::vector<double> supported_quantiles_ = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
```
const

ramaraochavali replied:
i tried making it const earlier. but if I make it, I need to use const_cast to pass to hist_approx_quantile method and after that I get bunch of errors like this
copy assignment operator of 'HistogramStatisticsImpl' is implicitly deleted because field 'supported_quantiles_' has no copy assignment operator

jmarantz replied:
If hist_approx_quantile method doesn't declare its read-only args const, perpahs @postwait could address that upstream?

And just go ahead and explicitly delete the copy-ctor and operator= methods: https://stackoverflow.com/questions/6077143/disable-copy-constructor

postwait replied:
Addressed upstream... the input `double *` for quantiles requested is now a `const double *`

ramaraochavali replied:
@jmarantz if I delete assignment operator, I need to assign it after merge like this
cumulative_statistics_ = HistogramStatisticsImpl(cumulative_histogram_);
Is there any alternative to it?

jmarantz replied:
Yes: I'd declare it as:
```
   std::unique_ptr<HistogramStatisticsImpl> cumulative_statististics_;
```
and then assign it as
```
   cumulative_statistics_ = std::make_unique<HistogramStatisticsImpl>(cumulative_histogram_);
```
that way you won't be copying any histograms by value.

ramaraochavali replied:
In the last review cycle I changed this from a ptr to heap allocation as @mattklein123 felt it is better that way. I can change it back to pointer though.

jmarantz replied:
I haven't 'suggested any change to how the vector<double> gets allocated, just how the class containing it gets managed.

ramaraochavali replied:
sure. agree. not in this class, in the place where this is used, it was pointer. No issues. I will change to pointer again.

jmarantz replied:
I'm not sure what comment of @mattklein123's you are referring to; if there is a disagreement go with his approach as I'm still learning this codebase myself; I'm just always paranoid about copying classes by value without scrutinizing that it's an appropriate operation, and one thing to be particularly careful of is member variables to objects managed by C libraries.

WRT this particular member var, I'm curious if a simple syntax change would allow you to copy-construct it as a const, declare the value as:
```
 const std::vector<double> supported_quantiles_{0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
```
I'm not sure if that will work or help as I'm actually new to some of the more modern constructs like initializing member variables where they are declared.  You could also get the copy-ctor to work by making an explicit one:
```
  HistogramStatisticsImpl(const HistogramStatisticsImpl& src)
    : supported_quantiles_(src.supported_quantiles_) {
  }
```
but you have to be diligent about auditing all the member vars and making sure it makes sense to copy them.  Move-semantics with std::unique_ptr on the class object is a lot easier.

Another model is to keep the HistogramStatisticsImpl allocated by value where its contained, as you have it now, and not re-assign it, but instead have a clear() method and re-populate it as needed.  That's actually the most efficient approach in that you don't have to copy that vector, though you may need to clear your accumulator values.

ramaraochavali replied:
I thought about this yesterday again. I ended up adding refresh method that takes the new histogram. It clears the existing one and recomputes again. So now there will be only Statistics object per histogram

jmarantz commented on source/common/stats/thread_local_store.cc:
```diff
@@ -75,6 +91,29 @@ void ThreadLocalStoreImpl::shutdownThreading() {
+
+void ThreadLocalStoreImpl::mergeInternal(PostMergeCb mergeCb) {
+  if (!shutting_down_) {
+    for (auto histogram : histograms()) {
```
explicit type here & other range loops where the type is not locally obvious.

ramaraochavali replied:
for range based loops on maps, looks like auto style c++ 11 is being used across the code. Do you want to change this here using iterator?

jmarantz replied:
Range-based for-loops are great.  Just asking you to be specific with the type of the loop var when not obvious, which IMO is the case here.  See the Google C++ style guide (applicable in Envoy with exceptions): https://google.github.io/styleguide/cppguide.html#auto)

There may be other cases in Envoy where 'auto' is used in lieu of a non-obvious type as well, but I'm only commenting on the code in this PR and the style guide.

One practical value of being explicit is that then I can see more clearly whether you are copying the object or referencing it, which is non-obvious with auto.

I am not asking you to use an explicit iterator object, though in another comment I noted that iterator variables usually have obvious type, and are thus appropriate for 'auto'.

ramaraochavali replied:
ok. Thanks. Got it. I changed this specific case.

jmarantz commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,145 @@
+private:
+  bool usedWorker() const {
+    bool any_tls_used = false;
+    for (TlsHistogramSharedPtr tls_histogram : tls_histograms_) {
```
Can you try:
```
  for (const TlsHistogramSharedPtr& tls_histogram : ...)
```
the benefit is that the shared-pointer won't have to inc/dec ref-counts during iteration, and the references are held through the loop by tls_histograms_.

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -333,27 +339,62 @@ class GaugeImpl : public Gauge, public MetricImpl {
+    for (double quantile : supported_quantiles_) {
+      summary_stream << "P" << quantile * 100 << ": " << computed_quantiles_[index];
+      if (index < supported_quantiles_.size() - 1) {
+        summary_stream << ", ";
```
much nicer.  WDYT of using absl::StrJoin and fmt::format as a slight simplification.

```
  std::vector<std::string> summary;
  for (size_t i= 0; i< supported_quantiles_.size(); ++i) {
    summary.push_back(fmt::format("P{}: {}", 100*supported_quantiles_[i], computed_quantiles_[i]);
  }
  return absl::StrJoin(summary, ", ");
```
5 lines instead of 10, fwiw.

jmarantz commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,145 @@
+    hist_free(histograms_[1]);
+  }
+  // Stats::Histogram
+  void recordValue(uint64_t value) override {
```
suggest grouping together all the overrides and only inlining them if they fit on one line.  Then you can omit the inter-line spacing for the overrides.

beginMerge() below is not an override.

ramaraochavali replied:
sorry i misread the comment. I thought you were just asking to move all overrides to one place and did not pay attention to "inline" word - will take care of it in my next commit.

jmarantz commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,145 @@
+    histograms_[1] = hist_alloc();
+  }
+
+  ~ThreadLocalHistogramImpl() {
```
virtual

jmarantz commented on test/common/stats/thread_local_store_test.cc:
```diff
@@ -190,6 +190,182 @@ TEST_F(StatsThreadLocalStoreTest, BasicScope) {
+                histogram->intervalStatistics().summary());
+      EXPECT_EQ(histogram->cumulativeStatistics().summary(), summary_with_one_as_value);
+    }
+  }
```
It's hard to see the structure in this long  test.  Consider factoring out common blocks as helpers, with any differences between them parameterized.  Or use a loop over structs with the iteration-specific values as members of the struct (golang style).

mattklein123 commented:
@ramaraochavali at this point I'm going to stop following the PR progress for a bit. Can you get past @jmarantz and @mrice32 reviews and then I will look again? @jmarantz @mrice32 @ramaraochavali  feel free to ping me if you have any questions in the interim. Thank you. This is going to be awesome.

ramaraochavali commented:
@jmarantz I have addressed all the comments except the test comments which I will do tomorrow when I add more tests. Can you PTAL and see if I have missed any?

jmarantz commented:
@ramaraochavali There's a handful of comments I've left that don't look like they are addressed yet to me. 

Can you take another pass over the 'Conversation' view, poking at the 'show outdated' links in GitHub?  

Then I'll take another look.  You can also DM me on Slack if you like and I can chat about C++ stuff if that helps.

ramaraochavali commented:
@jmarantz  I have looked at the conversations view and addressed all. Can you PTAL?  Please let me know I missed addressing any comment by mistake.
I have modified tests to be more readable, testing them will push that commit soon.

ramaraochavali commented on test/common/stats/thread_local_store_test.cc:
```diff
@@ -80,6 +80,63 @@ class StatsThreadLocalStoreTest : public testing::Test, public RawStatDataAlloca
+    for (const Stats::HistogramSharedPtr& histogram : histogram_list) {
+      if (histogram->name().find(hist_name) != std::string::npos) {
+        EXPECT_EQ(histogram->cumulativeStatistics().summary(), h1_cumulative_statistics.summary());
+        // EXPECT_EQ(histogram->intervalStatistics().summary(), h1_interval_statistics.summary());
```
i will enable this check based on @postwait reply on the issue I asked him.

postwait replied:
I can't find this communication.  Can you send this issue to me directly? Or file an issue on libcircllhist?

ramaraochavali replied:
@postwait Here is the issue
looks like found another corner case issue or may be i am misunderstanding.

Consider the code

histogram_t* main_thread_interval_hist = hist_alloc();
histogram_t* per_thread_interval_hist = hist_alloc();
hist_insert_intscale(per_thread_interval_hist, 1, 0, 1);
double in[9] = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
double out[9];
hist_approx_quantile(main_thread_interval_hist, in, 9,out);
histogram_t* hist_array[1];
hist_array[0] = per_thread_interval_hist;
hist_accumulate(main_thread_interval_hist, hist_array, ARRAY_SIZE(hist_array));
hist_clear(per_thread_interval_hist);
//hist_free(per_thread_interval_hist);
//per_thread_interval_hist = hist_alloc();
hist_array[0] = per_thread_interval_hist;
hist_insert_intscale(per_thread_interval_hist, 2, 0, 1);
//hist_clear(main_thread_interval_hist);
//hist_free(main_thread_interval_hist);
main_thread_interval_hist = hist_alloc();
hist_accumulate(main_thread_interval_hist, hist_array, ARRAY_SIZE(hist_array));
hist_approx_quantile(main_thread_interval_hist, in, 9,out);
std::cout<<"after accumulations :"<<out[0]<<"==> \n";

histogram_t* direct_hist = hist_alloc();
hist_insert_intscale(direct_hist, 2, 0, 1);
hist_approx_quantile(direct_hist, in, 9,out);
std::cout<<"direct :"<<out[0]<<"==> \n";

I expect direct and after accumulations to give the same output. But it gives 1 in first bin (rest of the bins are same) for accumulation case and 2 in the direct case.

If I change the code to free and hist_alloc it works correctly in both the cases. Is this is a corner case issue with hist_clear?

postwait replied:
Sorry confusing example of allocating without freeing.  Hard to follow the example.  I think you're right... There's an issue with empty bins.

postwait replied:
Fixed upstream.

ramaraochavali replied:
@postwait thanks a lot. Updated the lib and it works great.

ramaraochavali commented:
@jmarantz  @mattklein123  fyi, tested this manually today by running envoy and a test service continuously pushing requests for an hour. At the end of it the histograms are captured as follows 
```
cluster.test.upstream_rq_time: 
 	 Interval:   P0: 0, P25: 0, P50: 0, P75: 0, P90: 0, P95: 1.01228, P99: 1.08246, P99.9: 1.09825, P100: 1.1 
 	 Cumulative: P0: 0, P25: 0, P50: 0, P75: 0, P90: 1.08335, P95: 2.07604, P99: 29.9808, P99.9: 33.4218, P100: 120
```

ramaraochavali commented on include/envoy/stats/stats.h:
```diff
@@ -114,6 +115,11 @@ class Metric {
   virtual const std::string& tagExtractedName() const PURE;
+
+  /**
+   * Indicates whether a metric has been used.
```
this actually is a method used in other Metric as well. Used here signifies whether it has been updated/a value has been set because of some thing happening like connection closing etc. This is helpful in flushing only metrics whose value has been modified. I just moved it to the Metric interface now because histogram also uses it. Earlier Counter and Gauge has this method.   I am not sure if it is equivalent of empty.

ramaraochavali replied:
it is not since last flush, it indicates it has actually ever been set a value.

jmarantz replied:
No?  Then what is it used for?  Above you said "flushing only metrics whose value has been modified."

Over time, couldn't you wind up with every metric in this state, if the bit isn't cleared during flush?

ramaraochavali replied:
So it works like this. If a metric value has ever been set it is considered "used" and it would be pushed to stats sinks on subsequent flusher irrespective of whether it has been updated in that interval or not. I think it is reasonable because and would allow only metrics that are really used in that envoy instance to be pushed rather than all metrics. @mattklein123 may have better idea on the history. So I am not changing any thing here.

ramaraochavali commented on test/common/stats/thread_local_store_test.cc:
```diff
@@ -189,6 +246,251 @@ TEST_F(StatsThreadLocalStoreTest, BasicScope) {
 }
+TEST_F(StatsThreadLocalStoreTest, BasicSingleHistogramMerge) {
+  InSequence s;
+  store_->initializeThreading(main_thread_dispatcher_, tls_);
```
I thought about it and since the flow in each method/test case is little different, I felt it is difficult to factor them into a helper method.

jmarantz replied:
OK I'll try to point out some small helper-function opportunities that would make it easier to read.

ramaraochavali replied:
I have changed this.

jmarantz commented on test/common/stats/thread_local_store_test.cc:
```diff
@@ -189,6 +246,251 @@ TEST_F(StatsThreadLocalStoreTest, BasicScope) {
+  std::vector<uint64_t> h1_cumulative_values;
+  std::vector<uint64_t> h2_cumulative_values;
+  std::vector<uint64_t> h1_interval_values;
+  std::vector<uint64_t> h2_interval_values;
```
FWIW you could declare all these in one line.  Alternatively you could also make these vectors be class member vars, since you are using the same pattern in many tests.

jmarantz commented on test/common/stats/thread_local_store_test.cc:
```diff
@@ -189,6 +246,251 @@ TEST_F(StatsThreadLocalStoreTest, BasicScope) {
+  EXPECT_CALL(sink_, onHistogramComplete(Ref(h1), 0));
+  h1.recordValue(0);
+  h1_cumulative_values.push_back(0);
+  h1_interval_values.push_back(0);
```
I would turn this repeated 4-line stanza a helper method, e.g.

  expectCallAndAccmumute(expected_call, record_value, cumulative_value, interval_value) {
     ...
  }

...
  {
    expectCallAndAccmumute(43, 43, 43, 43);
    expectCallAndAccmumute(41, 41, 41, 41);
...

It's interesting that when you factor out all this repeated stuff you might see patterns more easily.  I imagine that in other sub-tests the 4 args will not all be the same?

ramaraochavali replied:
ok. Good pattern. I have even simplified further by 
`expectCallAndAccumulate(Histogram& histogram, uint64_t record_value)` as `record_value` is what we push to interval as well as cumulative histograms

jmarantz commented on test/common/stats/thread_local_store_test.cc:
```diff
@@ -80,6 +80,63 @@ class StatsThreadLocalStoreTest : public testing::Test, public RawStatDataAlloca
   }
 
+  void validateMerge(std::string hist_name, std::vector<uint64_t> h1_cumulative_values,
+                     std::vector<uint64_t> h2_cumulative_values,
```
if you are not going to make these member vars (suggested below) you should definitely pass them by const ref.

mattklein123 commented:
> I'm trying to understand how agnostic we want Envoy to be wrt to some of its dependencies. This binds Envoy's core source pretty tightly to circlehist, which might be the right thing in this case, but I know there are other histogram impls as well. Part of the broad appeal of Envoy I thought in general was the ability to integrate it into lots of different environments. So it'd be good to have @htuch and @mattklein123 and @alyssawilk weigh in on this aspect for this PR.

I think this is the right approach for this feature (after cleanup and code review). Although it's possible to do this as a sink, the histograms are generally useful for local debugging on the admin endpoint regardless of sink. Also, we will likely want this functionality later for latency based load balancing. 

Although there are multiple histogram implementations, after research I think it really comes down to this one and the HDR implementation which is more complicated. I think this implementation is pretty simple and a good place to start. If there is real pushback on having this feature in the core I would like to understand why. If perf is the concern, we can disable histogram collection via a stats store option. What else?

mrice32 commented:
@mattklein123 +1 to the idea of adding a stats store option to disable the histogram generation/collection.

ramaraochavali commented on source/common/stats/stats_impl.h:
```diff
@@ -333,27 +340,70 @@ class GaugeImpl : public Gauge, public MetricImpl {
+public:
+  HistogramStatisticsImpl() : computed_quantiles_(supported_quantiles_.size(), 0.0) {}
+
+  HistogramStatisticsImpl(histogram_t* histogram_ptr)
```
moved

ramaraochavali commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -91,6 +91,22 @@ void InstanceImpl::runOnAllThreads(Event::PostCb cb) {
   cb();
 }
 
+void InstanceImpl::runOnAllThreadsWithBarrier(Event::PostCb cb,
```
The 'barrier' was added to indicate that there is a barrier that waits for all threads to complete before calling the completion CB. If you have any better names I am open to changing this.

ramaraochavali commented on source/common/stats/thread_local_store.h:
```diff
@@ -16,6 +16,144 @@
+                           std::vector<Tag>&& tags)
+      : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent),
+        current_active_(0), flags_(0) {
+    histograms_[0] = hist_alloc();
```
I did not attempt to build the layer of abstraction as it is non trivial at this point and also the change is in single file which is easy to do if and when we want to move different histogram impl.

ramaraochavali commented:
@mattklein123 +1 to disabling it based on config. But would like to do it in a follow-up PR as this one is already huge and would like avoid further churn here.

ramaraochavali commented:
@jmarantz addressed all the comments. Whereever I have not addressed I have replied with a question/clarification. PTAL and it should be good now.

alyssawilk commented:
I'm going to go ahead and ping @mattrice for if we care to be able to
compile out the dependency.

My guess is we'd want to, and unfortunately starting with requiring it and
then removing the requirement with a follow-up would cause churn on our end
so [if I'm right and] if we can avoid it it'd be nice.  I'll let Matt
confirm though :-)

On Wed, Apr 11, 2018 at 3:16 AM, Rama Chavali <notifications@github.com>
wrote:

> @jmarantz <https://github.com/jmarantz> addressed all the comments.
> Whereever I have not addressed I have replied with a
> question/clarification. PTAL and it should be good now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/2932#issuecomment-380351917>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ARYFvR7D9QAQnXOWvrJbLCltCzsW6ps4ks5tna3OgaJpZM4TAQ35>
> .
>

mattrice commented:
@alyssawilk I think you probably wanted to ping @mrice32 instead of me. I do not feel I have the necessary  familiarity with this project to render an opinion :smile:

mattklein123 commented:
Unfortunately I can't even open this GH issue anymore. It unicorns for me
almost every time I'm guessing because it has too many comments. We may
need to close it and open a new one which references the old one.

Is it really necessary that this is compiled out right away? The library is
a single C file and header file. IMO this is going to cause a bunch of
extra friction on this PR which isn't really necessary. If compilation out
is required in the future I think the correct way to do it is:

1) Still have built in support within the stats store for histograms. I
think this is important.
2) Have the actual histogram implementation be compile time pluggable, with
a null implementation if the circle one is compiled out.

I don't think we need to do the full registration dance here.

On Wed, Apr 11, 2018 at 9:30 AM, alyssawilk <notifications@github.com>
wrote:

> I'm going to go ahead and ping @mattrice for if we care to be able to
> compile out the dependency.
>
> My guess is we'd want to, and unfortunately starting with requiring it and
> then removing the requirement with a follow-up would cause churn on our end
> so [if I'm right and] if we can avoid it it'd be nice. I'll let Matt
> confirm though :-)
>
> On Wed, Apr 11, 2018 at 3:16 AM, Rama Chavali <notifications@github.com>
> wrote:
>
> > @jmarantz <https://github.com/jmarantz> addressed all the comments.
> > Whereever I have not addressed I have replied with a
> > question/clarification. PTAL and it should be good now.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/envoyproxy/envoy/pull/2932#issuecomment-380351917>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> ARYFvR7D9QAQnXOWvrJbLCltCzsW6ps4ks5tna3OgaJpZM4TAQ35>
>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/2932#issuecomment-380514917>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGA17P_u332ZCAn9gOT-adOWqRJGXeUDks5tni-9gaJpZM4TAQ35>
> .
>



-- 
Matt Klein
Software Engineer
mklein@lyft.com
https://calendly.com/mattklein123

mattklein123 commented:
Rama,

I can't believe I'm saying this but can you close and open a new PR? Most
of us can't open the PR anymore. Please just reference the old PR in case
at some point in the future we can load it.

Thanks,
Matt

On Wed, Apr 11, 2018 at 9:35 AM, Matt Klein <mklein@lyft.com> wrote:

> Unfortunately I can't even open this GH issue anymore. It unicorns for me
> almost every time I'm guessing because it has too many comments. We may
> need to close it and open a new one which references the old one.
>
> Is it really necessary that this is compiled out right away? The library
> is a single C file and header file. IMO this is going to cause a bunch of
> extra friction on this PR which isn't really necessary. If compilation out
> is required in the future I think the correct way to do it is:
>
> 1) Still have built in support within the stats store for histograms. I
> think this is important.
> 2) Have the actual histogram implementation be compile time pluggable,
> with a null implementation if the circle one is compiled out.
>
> I don't think we need to do the full registration dance here.
>
> On Wed, Apr 11, 2018 at 9:30 AM, alyssawilk <notifications@github.com>
> wrote:
>
>> I'm going to go ahead and ping @mattrice for if we care to be able to
>> compile out the dependency.
>>
>> My guess is we'd want to, and unfortunately starting with requiring it and
>> then removing the requirement with a follow-up would cause churn on our
>> end
>> so [if I'm right and] if we can avoid it it'd be nice. I'll let Matt
>> confirm though :-)
>>
>> On Wed, Apr 11, 2018 at 3:16 AM, Rama Chavali <notifications@github.com>
>> wrote:
>>
>> > @jmarantz <https://github.com/jmarantz> addressed all the comments.
>> > Whereever I have not addressed I have replied with a
>> > question/clarification. PTAL and it should be good now.
>> >
>> > —
>> > You are receiving this because you were mentioned.
>> > Reply to this email directly, view it on GitHub
>> > <https://github.com/envoyproxy/envoy/pull/2932#issuecomment-380351917>,
>> > or mute the thread
>> > <https://github.com/notifications/unsubscribe-auth/ARYFvR7D9
>> QAQnXOWvrJbLCltCzsW6ps4ks5tna3OgaJpZM4TAQ35>
>>
>> > .
>> >
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/envoyproxy/envoy/pull/2932#issuecomment-380514917>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AGA17P_u332ZCAn9gOT-adOWqRJGXeUDks5tni-9gaJpZM4TAQ35>
>> .
>>
>
>
>
> --
> Matt Klein
> Software Engineer
> mklein@lyft.com
> https://calendly.com/mattklein123
>



-- 
Matt Klein
Software Engineer
mklein@lyft.com
https://calendly.com/mattklein123

ramaraochavali commented:
Closing this and will open another one
